sa/sa.go

```diff
@@ -2099,34 +2245,207 @@ func (ssa *SQLStorageAuthority) getChallengesImpl(db dbSelector, authID string)
 	if err != nil {
-		return 0, err
+		return nil, err
 	}
 	err = ssa.dbMap.Insert(am)
 	if err != nil {
-		return 0, err
+		return nil, err
+	}
+	return &sapb.AuthorizationID2{Id: &am.ID}, nil
+}
+
+// NewAuthorizations2 adds a set of new style authorizations to the database and returns
+// either the IDs of the authorizations or an error. It will only process corepb.Authorization
+// objects if the V2 field is set. This method is intended to deprecate AddPendingAuthorizations
+func (ssa *SQLStorageAuthority) NewAuthorizations2(ctx context.Context, req *sapb.AddPendingAuthorizationsRequest) (*sapb.AuthorizationIDs, error) {
+	ids := &sapb.AuthorizationIDs{}
+	for _, authz := range req.Authz {
+		id, err := ssa.NewAuthorization2(ctx, authz)
+		if err != nil {
+			return nil, err
+		}
+		ids.Ids = append(ids.Ids, fmt.Sprintf("%d", *id.Id))
 	}
-	return am.ID, nil
+	return ids, nil
 }
 
-// GetAuthz2 returns the authz2 style authorization identified by the provided ID or an error.
-// If no authorization is found matching the ID a berrors.NotFound type error is returned.
-func (ssa *SQLStorageAuthority) GetAuthz2(ctx context.Context, id *sapb.AuthorizationID2) (*corepb.Authorization, error) {
-	obj, err := ssa.dbMap.Get(authz2Model{}, id.Id)
+// GetAuthorization2 returns the authz2 style authorization identified by the provided ID or an error.
+// If no authorization is found matching the ID a berrors.NotFound type error is returned. This method
+// is intended to deprecate GetAuthorization.
+func (ssa *SQLStorageAuthority) GetAuthorization2(ctx context.Context, id *sapb.AuthorizationID2) (*corepb.Authorization, error) {
+	obj, err := ssa.dbMap.Get(authz2Model{}, *id.Id)
 	if err != nil {
 		return nil, err
 	}
 	if obj == nil {
-		return nil, berrors.NotFoundError("authorization %d not found", id)
+		return nil, berrors.NotFoundError("authorization %d not found", *id.Id)
 	}
 	return modelToAuthzPB(obj.(*authz2Model))
 }
 
+func authz2ModelMapToPB(m map[string]authz2Model) (*sapb.Authorizations, error) {
+	resp := &sapb.Authorizations{}
+	for k, v := range m {
+		// Make a copy of k because it will be reassigned with each loop.
+		kCopy := k
+		authzPB, err := modelToAuthzPB(&v)
+		if err != nil {
+			return nil, err
+		}
+		resp.Authz = append(resp.Authz, &sapb.Authorizations_MapElement{Domain: &kCopy, Authz: authzPB})
+	}
+	return resp, nil
+}
+
+// GetAuthorizations2 returns any valid or pending authorizations that exist for the list of domains
+// provided. If both a valid and pending authorization exist only the valid one will be returned. This
+// method is intended to deprecate GetAuthorizations.
+func (ssa *SQLStorageAuthority) GetAuthorizations2(ctx context.Context, req *sapb.GetAuthorizationsRequest) (*sapb.Authorizations, error) {
+	qmarks := make([]string, len(req.Domains))
+	for i := range req.Domains {
+		qmarks[i] = "?"
+	}
+
+	queryPrefix := fmt.Sprintf("SELECT %s FROM authz2", authz2Fields)
+	if *req.RequireV2Authzs {
```

Comment:
I think this flag isn't needed. AFAICT there's one call site for `SA.GetAuthorizations`, and it always specifies `RequireV2Authzs: true`. Also, I think the behavior we actually want is to *not* require v2 authzs (see below).

Also, is the intent to change the existing place where RA calls `SA.GetAuthorizations` so it calls `SA.GetAuthorizations2`?

Reply:
> I think this flag isn't needed. AFAICT there's one call site for SA.GetAuthorizations, and it always specifies RequireV2Authzs: true. Also, I think the behavior we actually want is to not require v2 authzs (see below).

Fine with me, I think we _do_ want to keep requiring v2 (in terms of the API, not the storage format) authorizations here though. I'm not sure what you are referring to when you say "see below"?

> Also, is the intent to change the existing place where RA calls SA.GetAuthorizations so it calls SA.GetAuthorizations2?

Not replace outright, it'd be a simple `if` along the lines of
```
if features.Enabled(features.NewAuthorizationSchema) {
	authzs, err = sa.GetAuthorizations2(ctx, req)
} else {
	authzs, err = sa.GetAuthorizations(ctx, req)
}
```

Reply:
Oh dang, I was worried we would get into trouble with the confusing terminology around v2 API vs v2 storage of authz. It looks like I confused those here. I'll take a second look with that in mind.

Reply:
Oh and the "see below" was a reference to https://github.com/letsencrypt/boulder/pull/4134#discussion_r274663243, where I was commenting on "	// TODO: In theory, we don't _need_ to do this. Instead we could"

Also, where is the call site that calls `sa.GetAuthorizations(... RequireV2Authzs: false)`?

Reply:
ah okay, so yeah not related because of the v2 api/v2 storage format mixup.

> Also, where is the call site that calls sa.GetAuthorizations(... RequireV2Authzs: false)?

There isn't one, I'm fine defaulting to the current behavior here, which is only returning v2 (API) authorizations.

Reply:
Yeah, I think having `SA.GetAuthorizations` return only ACMEv2 authorizations makes sense because it is only called in NewOrder. RA.NewAuthorization (the ACMEv1 path) calls `SA.GetValidAuthorizations`, which allows ACMEv1 authorizations (and will be replaced by `GetValidAuthorizations2`)